Add  file handle support for overlapped I/O

klu-dev opened the pull request:
Closes https://github.com/smol-rs/polling/issues/97

This is an implementation to support Windows file handle overlapped I/O. The basic idea is:

1. Add an File struct in PacketInner. The File struct have read and write OVERLAPPED struct. The read and write struct address will exposed to callers. The caller use read and write overlapped pointer as parameter for ReadFile/WriteFile operation.
2. IOCP completion events receive read and write address which will be converted back to Packet for updating events
3. New PollerIocpFileExt trait which has add_file and remove_file API to extend file functionality.

notgull reviewed (changes requested):
Thank you! I imagine this was a significant effort. I appreciate you writing this. However, I definitely need to read through this and understand this. So this review may take some time.

The immediate thing that jumps out at me is the weird pointer offsets in the `FileOverlapped` traits. This seems fragile to me. I would prefer them to be direct pointers; I think that would be better when it comes to strict provenance.

Also, have you run this code under Miri on Windows?

klu-dev commented:
Hi @notgull 

Thanks for reviewing. FileCompletionHandle trait would provide conversion from  overlapped pointer to packet which will depends on FileOverlapped trait providing pointer offset to packet. Maybe should rename FileOverlapped to FileOverlappedOffset.

I do not run the code with miri. Currently I am writing NamedPipe code in async_io crate. I found the code need some change in order for better integration with async_io. I also found a problem that is IOCP port still can receive event after the file handle is removed from the poller and file handle is closed as long as there is I/O operation with overlapped pointer. But the packet has been dropped at this time. This will cause memory access violation

I will request review after the issue is fixed. I will also run code in miri and solve your comments.

klu-dev commented on src/iocp/mod.rs:
```diff
@@ -427,6 +434,91 @@ impl Poller {
+    ///
+    /// IocpFilePacket will return both read and write overlapped pointer no matter what intrest events are.
+    /// The caller need to use the correct overlapped pointer for I/O operation. Such as: the read overlapped
+    /// pointer can be used for read operations, and the write overlapped pointer can be used for write operations.
```
Updated

klu-dev commented on src/iocp/port.rs:
```diff
@@ -292,8 +394,110 @@ impl<T: CompletionHandle> OverlappedEntry<T> {
+pub(super) struct CompletionKey(usize);
+
+static NEXT_DEFAULT_TOKEN: AtomicUsize = AtomicUsize::new(1); // 0 reserved for default iocp packet
+static NEXT_FILE_TOKEN: AtomicUsize = AtomicUsize::new(1usize << (usize::BITS - 1)); // Initialize with high bit set
```
Yes. Add a new CompletionKeyGenerator object owned by every IoCompletionPort. Different port can have the same completion key.

arnodb commented:
Hi, first of all thanks for the hard work!

I don't know if this PR is still on the radar, but I am experimenting with it and the `async-io` paired branch. My code is located at:

https://github.com/arnodb/teleop/tree/windows-named-pipe

My results so far are visible at https://github.com/arnodb/teleop/actions/runs/21923871658 (or more recent runs of CI on the branch).

Surprisingly the unit test `attach::windows_named_pipe::tests::test_named_pipe_attachment` passes as opposed to the real life example which:

* seems to receive multiple client connections (according to the debug I added)
* seems to panic in the client (`Task polled after completion`) but it may be cause by another error I am not aware of (it could also be an issue in `async-io` when things go wrong)

I hope this use case can help making progress on overlapped I/O in `polling` and `async-io`.

arnodb commented:
@klu-dev, `Task polled after completion` is simply a `self.flush = None;` missing in `poll_flush` (`async-io` branch). I'm now investigating other issues.

klu-dev commented:
@arnodb Sorry for reply lately. I am looking at the issue. The crash comes from the poll_flush can be fixed easily. But I found another issue for implementation in async-io about windows  named pipe. The problem is named pipe events are edge triggered. The event is only triggered once, so current implmentation does not support poll many times like  select! macro in the server.rs in teleop.  I am trying to fix this issue.

klu-dev commented:
@arnodb The issue has been fixed in https://github.com/klu-dev/async-io/commit/fd02db0933ac79274ebc571b1c0a4b4c401bf215 . async-io repo is updated, but there is no code change for polling repo. I found another issue for your server.rs code, the future `stream = futures::StreamExt::next(&mut conn_stream).fuse() =>` will frequently be polled after there is client connection to server. The reason task runnner LocalPool do not wake up specific task independtly, instead it wake up the thread, and loop all task for poll, which cause any task task wake up will cause all tasks polled once.

async-io NamedPipeListener::bind() add PipeMode and NamedPipeOpenOptions parameters to support set pipe params. Your code need to change to `NamedPipeListener::bind(named_pipe_name(std::process::id()), Default::default(), None)?;` in windows_named_pipe.rs::listen code.

arnodb commented:
Hi @klu-dev, thanks for this update!

I pushed (force) the latest changes to my branch. It looks like it works now :tada:.

Regarding the `LocalPool` executor comment, I assume this is not an issue in my code strictly speaking but a tradeoff of the executor itself.

If I understand your comment correctly: all the tasks in the executor are polled whenever there is activity in one of them (e.g. a byte sent through the wire), correct?

I didn't know that (still new to `async` details), thanks. Is it an issue? I guess generally "not really", but in combination with your named pipes? Did you notice problematic behaviour?

(Good news: `teleop` does not force anyone to use `LocalPool` :sweat_smile:)

klu-dev commented:
@arnodb I found the root cause which is poll_flush bug in async-io for named pipe. It is fixed in commit https://github.com/klu-dev/async-io/commit/10126a1c639b5d0511d301db7d9a22e08351c8b3.

notgull commented:
Sorry for the delay; this is large PR and I haven't found the time to sit down and review the entire thing in one sitting.

klu-dev commented:
@notgull I update the code based on AI code review. The code improvement could refer to section 2 and 3 in `docs/named-pipe.design.md` doc

klu-dev committed:
Add file handle support for overlapped I/O

Adds Windows file-handle (named pipe / file / device) support to the IOCP
backend via a new completion-based API:

- New `PollerIocpFileExt` trait with `register_file` for binding any
  `FILE_FLAG_OVERLAPPED` handle to the poller.
- `RegisteredFile` exposes `submit_read`, `submit_write`, and
  `submit_connect_named_pipe`, returning a `Submission`
  (Complete / Pending / Failed).
- Pending submissions yield an `OpHandle<B>` that owns the operation
  lifecycle: completion state, NTSTATUS, bytes transferred, and buffer
  hand-back via `take`. The dispatcher emits
  `Event { key, readable, writable }` so reactors (e.g. async-io) keep
  ownership of waker registration.
- `set_user_key` lets reactors rebind the event key after registration.
- `StableBuf` / `StableBufMut` traits define the owned-buffer contract;
  blanket impls for `Vec<u8>`, `Box<[u8]>`, and `&'static [u8]`.
- IOCP completion key now uses a high-bit tag to discriminate socket vs
  file ops; per-op packets are reclaimed via `Arc::from_raw` in the
  dispatcher.
- NTSTATUS -> io::Error mapping covers ACCESS_DENIED, INVALID_HANDLE,
  INVALID_PARAMETER, BUFFER_TOO_SMALL, BUFFER_OVERFLOW, TIMEOUT, and
  IO_TIMEOUT; STATUS_BUFFER_OVERFLOW is treated as a successful short
  read.
- `submit_read` / `submit_write` reject buffers larger than u32::MAX up
  front.
- Bumps MSRV to 1.77 (required by `std::mem::offset_of!`).
- Adds design doc (`docs/named-pipe.design.md`), examples, and
  integration tests covering lifetime, concurrent multi-op,
  oversized-buffer rejection, key rebinding, and direction emission.

Closes the work tracked in PR #248.

klu-dev commented:
@notgull I changed the code from readiness mode to complete IO mode in order to solve resource safety challenge. The  design doc refer to https://github.com/klu-dev/polling/blob/kail/windows_file/docs/named-pipe.design.md. async-io for NamedPipe refer to https://github.com/klu-dev/async-io/blob/kail/windows_named_pipe/docs/named-pipe.design.md.
